### PR TITLE
[DS-4947] Add negative user sampling to evaluation

### PR DESCRIPTION
**Pull request by emmafgrimes**
## Describe the big picture - why are we doing this?
This is the (hopefully) final PR in adding the ability to negative sample users in `collie` models! Specifically, this is the PR that adds the ability to `metrics.py`. I also caught a couple other places in `utils.py` and `cross_validation.py` that needed updating as well. Please take your time reviewing this PR, it was the most difficult of the 3 for me. 
 
## Any additional details to clarify code in the PR?
The metrics code can be hard to follow at times, so as always, I am open to feedback in enhancing the readability with any refactoring.

## How is this tested?
On an EC2 instance all the tests passed. We have one final ticket for this work where I will rerun all the tutorial notebooks with negative user sampling. 

## Pull Request Checklist
 - [x] Pull request includes a description of why we are doing this
 - [x] ``CHANGELOG`` has been updated
 - [x] Version in ``_version.py`` has been updated
 - [x] All tests in the ``tests`` folder pass with a local build
 - [ ] ``README`` has been updated (if applicable)
 - [ ] Documentation in ``docs`` has been updated (if applicable)
 - [ ] ``requirements.txt`` and ``requirements-dev.txt`` have been recompiled (if applicable)
 - [x] Docker image can be built using ``docker build -t collie .``

### CODE REVIEW

File: collie/cross_validation.py
Comment (nathancooperjones) [inline collie/cross_validation.py]: I think we can simplify this logic a bit more: 

```suggestion
    users_or_items = interactions.mat.row
    
    if getattr(interactions, 'negative_sample_type') == 'user':
        users_or_items = interactions.mat.col
```

File: collie/cross_validation.py
Comment (nathancooperjones) [inline collie/cross_validation.py]: Could we try just passing in `opposite_type` to each worker in the parent function above so we don't have to re-compute this each time.

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: Same comment here - can we just pass in `negative_sample_type` here to avoid re-computing this each time?

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: This is a time when I wish we had a Nicole around. For now, I think this works, but before we think about merging into the main Collie codebase, I think we should try to establish a better naming scheme that makes the code more readable while still understanding that things can refer to either users or items. I hate that we have to put `user_or_item` everywhere now, but I have no better ideas.

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: What do the names `aabb` and `cdcd` mean?

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: We also need to update the `Example` section of the docs below this.

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: We can reduce repeated code here a bit more: 

```suggestion
    if negative_sample_type == 'item':
        labels = targets[user_or_item_ids[:, None], np.array(preds.detach().cpu())]
    elif negative_sample_type == 'user':
        labels = targets[np.array(preds.detach().cpu()), user_or_item_ids[:, None]]

    return torch.tensor(
        (labels > 0)
        .astype('double')
        .toarray(),
        requires_grad=False,
        device=device,
    )
```

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: I think we can also refactor this a bit more to reduce repeated code.

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: [Optional] I realize the original code wasn't written this way, but my new coding philosophy is to be explicit with keyword arguments as much as possible. I think it's a lot clearer (and less error-prone) to write function calls like this: 

```python
foo(a=a, b=b, c=c, d=d)
```

instead of 

```python
foo(a, b, c, d)
```

If you have energy and want to, I think it would be nice to do that for all of our internal function calls in this file. If not, feel free to ignore.

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: Ditto here with less repeated code.

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: Ditto here.

File: tests/fixtures/metrics_fixtures.py
Comment (nathancooperjones) [inline tests/fixtures/metrics_fixtures.py]: It looks like this is just `test_implicit_predicted_scores` but transposed. If that's intentional, can we just use that logic to make this fixture, rather than have it fully typed out?

File: tests/test_metrics.py
Comment (nathancooperjones) [inline tests/test_metrics.py]: Do we need to test the user-based version of this as well?

File: tests/fixtures/metrics_fixtures.py
Comment (nathancooperjones) [inline tests/fixtures/metrics_fixtures.py]: Why make this a separate fixture and not just update the original `targets` fixture?

File: tests/test_metrics.py
Comment (nathancooperjones) [inline tests/test_metrics.py]: Should we test `test_evaluate_in_batches` for user-based negative sampling as well?

File: tests/test_utils.py
Comment (nathancooperjones) [inline tests/test_utils.py]: I actually think we don't need to add this test, since we just use the column name directly. The above test already tests 100% of the code.

File: CHANGELOG.md
Comment (nathancooperjones) [inline CHANGELOG.md]: Typo. 

```suggestion
 - ``BasePipeline`` inherits  ``negative_sample_type`` from the ``train_loader``
```

File: CHANGELOG.md
Comment (nathancooperjones) [inline CHANGELOG.md]: Typo. 

```suggestion
 - ``negative_sample_type`` arguments in ``mapk``, ``mrr``, and ``auc`` and the calculations are computed accordingly
```

Comment (nathancooperjones) [reply CHANGELOG.md]: I'm also not sure we need to add the `and the calculations are computed accordingly` there since I think that is implied.

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
I think this is good work for V1 of CF to deploy at TMP, but for merging back into the main Collie repo, I think we got to talk about the right way to name and document all of this in as little a confusing way as possible.

File: collie/metrics.py
Comment (emmafgrimes) [reply collie/metrics.py]: Per offline discussion, `rows` and `cols` FTW!

File: collie/metrics.py
Comment (emmafgrimes) [reply collie/metrics.py]: It was the pattern of the output tensor, all has changed now!

File: collie/metrics.py
Comment (emmafgrimes) [reply collie/metrics.py]: I think I covered all the cases of this, but please call out any I missed in the next review - happy to edit!

File: collie/cross_validation.py
Comment (emmafgrimes) [reply collie/cross_validation.py]: This raises an `AttributeError` I added a default `if getattr(interactions, 'negative_sample_type', 'item') == 'user':` let me know what you think in the next review.

File: collie/cross_validation.py
Comment (nathancooperjones) [inline collie/cross_validation.py]: Should we do the same `row/col` naming scheme here too? My preference is to eliminate as many `user_or_item` variable names as possible.

File: collie/cross_validation.py
Comment (nathancooperjones) [inline collie/cross_validation.py:307]: Why does this function take `opposite_type` as an input but then recompute it in the line below?

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py]: Maybe make this example a little clearer with comments: 

```suggestion
        >>> # let's assume an interactions matrix with users as rows and item as columns
        >>> row_ids = np.array([10, 11, 12])  # user IDs 
        >>> n_cols = 4  # number of items
```

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py:72]: This is so much clearer!

File: collie/metrics.py
Comment (nathancooperjones) [inline collie/metrics.py:28]: Should we also change the description of this variable on the next line to not say `items or users` and just be consistent with `row/col`?

Comment (nathancooperjones) [reply collie/metrics.py:28]: Ditto to a lot of docstrings in this file.

File: tests/test_metrics.py
Comment (nathancooperjones) [inline tests/test_metrics.py]: Why `12` here? Is there a way to make this more dynamic so in case we change the fixture in the future, the test doesn't have to also be changed?

File: collie/cross_validation.py
Comment (nathancooperjones) [inline collie/cross_validation.py:153]: I think part of the `row/col` discussion was to make this consistent across all Collie files where we use `user or item`. What do you think about applying that to other files, like this one?

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
This looks _sooo_ much better, nice work! I think this initial work to move towards `row/col` is great, and I'm hoping we can do this in more files/functions. 

Also, I think we should hold off on merging until we address why the tests are taking so long to run.

File: collie/cross_validation.py
Comment (emmafgrimes) [reply collie/cross_validation.py:307]: OOPS!

File: collie/metrics.py
Comment (emmafgrimes) [reply collie/metrics.py:28]: Oh yes, apologies I meant to do that before submitting!

File: tests/test_metrics.py
Comment (emmafgrimes) [reply tests/test_metrics.py]: Unfortunately, `.view(-1)` doesn't work on the transposed tensor like in the test above. I can replace `12` with `torch.numel` tho:
```suggestion
    model.return_value = test_implicit_predicted_scores_negative_sample_user.reshape(
      torch.numel(test_implicit_predicted_scores_negative_sample_user)
    )
```

File: collie/cross_validation.py
Comment (emmafgrimes) [reply collie/cross_validation.py:153]: I am kind of struggling with when to change to `row/col`. For this specific file, I think `user or item` makes more sense since we didn't change the `datasets.py` file and `Interactions` objects are still defined as rows = users and cols = items up to this point. The `row/col` switch only really happens in metrics. Happy to discuss this further.

File: tests/test_metrics.py
Comment (nathancooperjones) [reply tests/test_metrics.py]: Love this!

File: collie/cross_validation.py
Comment (nathancooperjones) [reply collie/cross_validation.py:153]: That makes total sense! Just for the sake of improved clarity, in these docstrings, can we say something along the lines of ```user or item, depending on ``negative_sample_type`` ``` just to make it clear when one applies. 

Maybe even something like: 

```
user or item, depending on ``negative_sample_type`` (user when ``negative_sample_type == 'item'``, item when user when ``negative_sample_type == 'user'``) 
```

although this might be overkill.

File: tests/conftest.py
Comment (nathancooperjones) [inline tests/conftest.py]: We'll probably want to un-comment this out.

File: tests/test_model.py
Comment (nathancooperjones) [inline tests/test_model.py]: Same here.

File: tests/conftest.py
Comment (emmafgrimes) [reply tests/conftest.py]: Excuse me I did not request re-review yet ;p

File: collie/cross_validation.py
Comment (nathancooperjones) [inline collie/cross_validation.py:151]: Can we add this language to: 

* `ApproximateNegativeSamplingInteractionsDataLoader` L191 
* `Interactions` L262 
* `BasePipeline.calculate_loss` L591

**Comment by emmafgrimes** [comment]
Keeping an eye on the tests, but this is otherwise ready for re-review.

**Review by nathancooperjones — CHANGES_REQUESTED** [review]
Changes look good! But, the tests are still taking longer to run on some versions of Python. Can we investigate why this is happening before merging in?

**Comment by emmafgrimes** [comment]
> Changes look good! But, the tests are still taking longer to run on some versions of Python. Can we investigate why this is happening before merging in?

Yep! Doing that now...

**Comment by emmafgrimes** [comment]
WOO tests fixed! Thanks for the help @nathancooperjones , back to you!

**Review by nathancooperjones — APPROVED** [review]
LGTM! 🎉 congrats!!

**Pull request merged**
